File: praxis/2020/jun/2max.py
def sum_number(n):
    total = 0
    for d in str(n):
        total = total + int(d)
    return total

def max_lookup_dict(number_list):
    lookup_dict = {}
    for n in number_list:
        index = sum_number(n)
        if not index in lookup_dict:
            lookup_dict[index] = [n]
        elif n > lookup_dict[index][0]:
            lookup_dict[index][1] = lookup_dict[index][0]
            lookup_dict[index][0] = n
        elif len(lookup_dict[index]) == 1:
            lookup_dict[index].append(n)
        elif n > lookup_dict[index][1]:
            lookup_dict[index][1] = n
    return lookup_dict

def main():
    number_list = [51, 71, 17, 26, 12, 42]
    lookup_dict = max_lookup_dict(number_list)
    for key, values in lookup_dict.items():
        total = (-1, sum(values))[len(values) == 2]
        print(key, total, values)

main()

File: praxis/2020/jun/2max.rb
def sum_number(n)
  total = 0
  n.to_s.split("").each do |c|
    total += c.to_i
  end
  total
end

def lookup_max_list(number_array)
  lookup_list = {}
  number_array.each do |n|
    sum = sum_number(n)
    if lookup_list[sum].nil?
      lookup_list[sum] = [n]
    elsif lookup_list[sum][1].nil? || n > lookup_list[sum][1]
      lookup_list[sum][1] = n
    end
  end

  lookup_list
end

list = [51, 71, 17, 26, 12, 42]
lookup_max_list(list).map do |key, values|
  total = (values.length === 2) ? values.sum() : -1
  puts "#{key}: #{total} [#{values.join(', ')}]"
end